### PR TITLE
[MRG] add ZipStorage, support loading tree from storage

### LINKED ISSUE #490: Support all-in-one SBT indexes
The default way of distributing sourmash indexes is thru a `tar.gz` that need to be uncompressed to be used (and contains the `sbt.json` index metadata and the actual nodes data inside a hidden dir). This is annoying because you need at least double the storage space to be able to hold the original `tar.gz` file and also the uncompressed data which will actually be used.

A better approach would be to take the same `tar.gz` and read the `sbt.json` file from it, and also access the actual data without uncompressing it (in disk, still need to uncompress it to memory). This can be done using the `tarfile` module, and we already have examples of something very close to this in the `TarStorage` class.

The difference between this solution and the `TarStorage` is only where the metadata is present: in the same archive, or in two separate files (one for metadata, other for node data)

@scanon and @psdehal suggested using the [zipfile](https://pymotw.com/3/zipfile/index.html) module because ZIPs have an index and support random access without needing to scan the full file.

### PR DESCRIPTION
**Pull request by luizirber**
Fixes #490 , closes #60 

## Checklist

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by luizirber** [comment]
Needs more testing, but it is slower than uncompressing the file and loading from the hidden dir for now

**Comment by luizirber** [comment]
Still missing:

- [ ] saving into a zipfile... I created `tests/test-data/v4.zip` manually, running
```
cd tests/test-data
zip -9 -rj v4.zip .sbt.v3
cp v4.sbt.json tree.sbt.json
zip -9 v4.zip tree.sbt.json
```

**Comment by luizirber** [comment]
From https://github.com/dib-lab/sourmash/pull/799#issuecomment-618561836:
> SBTs are currently v4; saving compressed nodes imply a v5 (because if we keep the same version number, older versions of sourmash can't load a v4 anymore if it is compressed...)
>
> There is already experimental support for loading v5 in #694, but we still save as v4 by default. Since we don't save as v5 yet, there is no officially supported v5 SBTs in existence, so it's OK to update the format (probably also adding some field for the LocalizedSBT in #925?)
>
> (Currently the difference from v4 to v5 is splitting leaves and internal nodes, which allows saving LinearIndex in the same format. Also, only the leaves are required, and maybe start calling them datasets or signatures is also good)

**Comment by luizirber** [comment]
Loading from a zipfile already works, and I implemented the same logic for tar files but it was waaay to slow (I tried with the current prepared DBs and... it just never finishes), so I removed it.

Following what we did with previous SBT updates, I propose:
- start saving to v5 (we support loading from previous versions, but always save to the most recent one)
- v5 prepared DBs are zip files, instead of tar files (so people don't try to run older DBs with newer sourmash versions)
- use zip files only as a containers, and store data as it is added to the containers. This avoids slowness in the `zipfile` Python module (since it will just outputs what was added, without trying to decompress)
- `core` is ready for loading compressed files (nodegraphs or signatures), so...
- start always saving nodegraphs as compressed. This is not compatible with `khmer` (only CountGraph can be gzipped), but... it doesn't matter?
- also compress signatures when they are added to an SBT storage (but not signatures in general, because it would break previous sourmash versions)

It is a bit weird to use the zip file just as containers, and do the compression/decompression outside it, but the main benefit is that even if the file is unzipped the `FSStorage` will use about the same size (because each individual node data will still be compressed).

**Comment by ctb** [comment]
👍

**Comment by luizirber** [comment]
OK, time to bikeshed: I'm checking for `.zip` extensions for loading/saving indices, but should it be `.sbt.zip` instead? I think I prefer the latter because it has `.sbt` in the name.

**Comment by ctb** [comment]
sure :)

On Mon, Apr 27, 2020 at 11:46:14AM -0700, Luiz Irber wrote:
> OK, time to bikeshed: I'm checking for `.zip` extensions for loading/saving indices, but should it be `.sbt.zip` instead? I think I prefer the latter because it has `.sbt` in the name.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/sourmash/pull/648#issuecomment-620164670
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by ctb** [comment]
I got this to work manually by taking an already-existing SBT and zipping it up. Nice!

Misc questions in initial conceptual review --

* what requirements are there for the name of the index? does it just find the first .sbt.json file in there?
* how do you create these with sourmash? Is this PR only adding the functionality to read them? Skimming the code I would have expected to be able to provide a '.sbt.zip' filename and automatically produce a zipfile, but it doesn't seem to work. That might be because the `sourmash index` CLI automatically adds a `.sbt.json` on to the end of files... if so, should we fix that here or in another, later PR? (That later PR could also add documentation :)
* this PR also seems to add loading of compressed (.gz) signatures!? ref #60

**Comment by luizirber** [comment]
> I got this to work manually by taking an already-existing SBT and zipping it up. Nice!

This is a nice side benefit, but it is slower. If you zip it up this way it's up to the Python `ZipFile` to decompress, and that is slower than letting the Rust code decompress. Hence why all the discussion of storing "uncompressed" zip files, because this way `ZipFile` will read the data and only when it is passed to Rust it will be properly decompressed.

> Misc questions in initial conceptual review --
> 
>  what requirements are there for the name of the index? does it just find the first .sbt.json file in there?

It find the ONLY `.sbt.json` in there. If there is more than one it complains at the moment:
https://github.com/dib-lab/sourmash/pull/648/files#diff-f01fb6a1122709b5d46828f26c23854aR609-R613

>  how do you create these with sourmash? Is this PR only adding the functionality to read them? Skimming the code I would have expected to be able to provide a '.sbt.zip' filename and automatically produce a zipfile, but it doesn't seem to work. That might be because the `sourmash index` CLI automatically adds a `.sbt.json` on to the end of files...

Was this before https://github.com/dib-lab/sourmash/pull/648/commits/667c95dd51fab49bc87cde9a44c9c598174e5022? The name was just `.zip` (instead of `.sbt.zip`) before this commit.

Up for discussion: default the index creation to zipped SBTs, if the name doesn't specify `.sbt.json`? For now it still defaults to `FSStorage` (the hidden dir).

Also important: SBTs created with whatever this version of sourmash ends up being (`3.3`?) won't be backward-compatible (because they will be saved as `v5`). So I think defaulting to zipped SBTs is OK.

> if so, should we fix that here or in another, later PR? (That later PR could also add documentation :)

It's in the TODO list, together with rewriting the docs to use the new trees. Maybe add the zipped SBTs creation to this PR, but punt tutorials/docs changes to another PR (but pre `3.3`?)

> this PR also seems to add loading of compressed (.gz) signatures!? ref #60

yup =]

They are still saved uncompressed by default, but the `SBT.save` method uses `compressed=True` in `save_signatures`, so they are compressed starting in `v5`.

**Comment by ctb** [comment]
- yep, creation with a `.sbt.zip` suffix now works!
- defaulting to zipped SBT seems like a breach of semantic versioning for v3, since the output filename(s) change, right? There *is* a simple way to override it - looks like explicitly specifying `.sbt.json` works ok. I dunno. Seems like having it be an easter egg for people until v4 is safest.
- good to bump to v5 by default now, I think? what happens if you try to load a v5 with v3.2?
- I think having a separate PR for docs and command line (keeping this PR from getting bigger and uglier) is better. That leaves this as "technical support for ZipStorage."

I still want to go through some of the bits of code and tests and understand them, tho.

**Comment by ctb** [comment]
(have you run this on a really big SBT? How well does it perform?)

**Comment by luizirber** [comment]
> (have you run this on a really big SBT? How well does it perform?)

## Indexing

Regenerating the prepared genbank SBT (94976 signatures).

`$ /usr/bin/time -v -- sourmash index -k 51 --traverse-directory genbank-k51.sbt.zip .sbt.genbank-k51`

32m12s, 8.3GB of RAM, 4 GB on disk.

<details>
```                                                                                                                                                  
== This is sourmash version 3.2.4.dev9+g667c95d.d20200427. ==                                                                                      
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==                                                                                 
                                                                                                                                                   
loading 94976 files into SBT                                                                                                                       
                                                                                                                                                                                                                                                                                                       
loaded 94976 sigs; saving SBT under "genbank-k51.sbt.zip"                                                                                          
                                                                         
Finished saving nodes, now saving SBT json file.
        Command being timed: "sourmash index -k 51 --traverse-directory genbank-k51.sbt.zip .sbt.genbank-k51"                                     
        User time (seconds): 1906.14                                     
        System time (seconds): 12.67                                     
        Percent of CPU this job got: 99%                                                                                                           
        Elapsed (wall clock) time (h:mm:ss or m:ss): 32:12.94            
        Average shared text size (kbytes): 0                                                                                                       
        Average unshared data size (kbytes): 0                                                                                                     
        Average stack size (kbytes): 0                                                                                                             
        Average total size (kbytes): 0                                                                                                             
        Maximum resident set size (kbytes): 8365948                                                                                                
        Average resident set size (kbytes): 0                                                                                                      
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 2148786
        Voluntary context switches: 67062
        Involuntary context switches: 6372
        Swaps: 0                                                         
        File system inputs: 11580531
        File system outputs: 7820025                                     
        Socket messages sent: 0                                          
        Socket messages received: 0                                      
        Signals delivered: 0                                             
        Page size (bytes): 4096                                          
        Exit status: 0
```
</details>

## Gather

`$ /usr/bin/time -v -- sourmash gather -k 51 HSMA33OT.fastq.gz.sig genbank-k51.sbt.zip `

2m53s, 2.17 GB of RAM.

<details>
```                                                                                                                                                
== This is sourmash version 3.2.4.dev9+g667c95d.d20200427. ==                                                                                      
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==                                                                                 
                                                                                                                                                   
selecting specified query k=51                                                                                                                     
loaded query: CAVL1ANXX170419:7:1101:10005:9... (k=51, DNA)                                                                                        
loaded 1 databases.         

[...]

found 59 matches total;                                                                                                                            
the recovered matches hit 62.4% of the query                                                                                                       
                                                                                                                                                   
        Command being timed: "sourmash gather -k 51 /home/luizirber/work/sourmash-bio/sourmash_resources_debug/HSMA33OT.fastq.gz.sig genbank-k51.sbt.zip"                                                                                                                                              
        User time (seconds): 172.45                                                                                                                
        System time (seconds): 2.22                                      
        Percent of CPU this job got: 100%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:53.16
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2169964
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 568322
        Voluntary context switches: 38
        Involuntary context switches: 236737
        Swaps: 0                                                         
        File system inputs: 6409914                                      
        File system outputs: 46949                                       
        Socket messages sent: 0                                          
        Socket messages received: 0                                      
        Signals delivered: 0                                             
        Page size (bytes): 4096                                          
        Exit status: 0    
```
</details>

**Comment by ctb** [comment]
re benchmarks, that's great - well worth it given the time needed to unpack the .tar.gz file! The zip file is only 4 GB!? That's amazing!

I did a line-by-line review and added some comments in #968. I'd suggest merge & then edit, but feel free to edit that branch directly, too.

Other issues to think about before a merge --

- [ ] make sure error message when trying to load zip file with old sourmash (2.x?) is useful.
- [ ] need a test for creation of zip file at command line

**Comment by ctb** [comment]
Built `gtdb-release89-k31.sbt.zip` in `/home/ctbrown/sourmash_databases/gtdb/build`, works great! 1.4 GB zip file.

**Comment by luizirber** [comment]
> defaulting to zipped SBT seems like a breach of semantic versioning for v3, since the output filename(s) change, right? There _is_ a simple way to override it - looks like explicitly specifying `.sbt.json` works ok. I dunno. Seems like having it be an easter egg for people until v4 is safest.

Makes sense

>  good to bump to v5 by default now, I think? what happens if you try to load a v5 with sourmash 3.2?

It will fail, but in a weird way. The `v5` JSON in 3.2 is different (`leaves` instead of `signatures`, no `index_type`). Maybe go straight into `v6` (and there is [historical precedent](https://itsfoss.com/what-happened-to-ipv5/) for that...)

And a reminder that we always support **loading** previous versions, but always **save** to the latest one. So, an SBT generated in `3.3` can't be used in `3.2`. We are backwards-compatible, but I don't think supporting whatever we develop in the future is a good strategy.

> I think having a separate PR for docs and command line (keeping this PR from getting bigger and uglier) is better. That leaves this as "technical support for ZipStorage."

:+1:

**Comment by luizirber** [comment]
> re benchmarks, that's great - well worth it given the time needed to unpack the .tar.gz file! The zip file is only 4 GB!? That's amazing!

We pay the price on index building, tho. I ran with current master and it took ~8 minutes (compared to 32 minutes for zipped), but I think it's better to pay the price when indexing than every time it is used.

(And I'm pretty sure the `ZipStorage` can be implemented in Rust and be faster, but won't do this now)

> I did a line-by-line review and added some comments in #968. I'd suggest merge & then edit, but feel free to edit that branch directly, too.

Thanks, will check.

> Other issues to think about before a merge --
> 
> make sure error message when trying to load zip file with old sourmash (2.x?) is useful.

Hmm, how? Will we release new 2.x versions?

> maybe modify CLI message "Finished saving nodes, now saving SBT json file." to add something about output location, and to differentiate b/t zip and other storage methods?

:+1:

**Comment by luizirber** [comment]
> We pay the price on index building, tho. I ran with current master and it took ~8 minutes (compared to 32 minutes for zipped), but I think it's better to pay the price when indexing than every time it is used.
> 
> (And I'm pretty sure the `ZipStorage` can be implemented in Rust and be faster, but won't do this now)

Another interesting tradeoff: turns out that level `1` (equivalent to `gzip -1`) is not that much slower, and has reasonable size. I was running with `9`, and it is a pretty big performance impact: almost 7x slower than no compression `0`, and 5x slower than `1`. All for a meager 6MB difference (from 252 MB in `1` to 246 MB in `9`).

Level | Size | Time
--- | --- | ---
0 | 407 MB | 16s
1 | 252 MB | 21s
5 | 250 MB | 39s
9 | 246 MB | 1m48s

(Decompression is all about the same time, in a quick test with `gather` all of them take 8s).

Tempted to default compression to `1`, and maybe expose to the `sourmash index` CLI options to change the compression level (this might be important space savings for the prepared databases, for example).

**Comment by ctb** [comment]
Did some dumpster diving back to sourmash 3.2 and 2.1.

tl;dr I think there are these TODO items:

- [ ] for the future, when trying to load an unsupported version of a database, emit an error message like "The database format in file <blahblah> is not supported this version of sourmash." rather than a KeyError
- [ ] I'm on board with bumping to v6, since the error message for v5 format change is bad :). In the release or commit notes let's make it clear somewhere that no v5 databases should exist in the wild...! We could wait until #925 is merged to do the bump, since that will probably be another incompatibility in created databases, right?
- [ ] we should never be in the situation where older versions of sourmash use khmer to try to load a compressed internal node, right?
- [ ] when creating a zip file with `sourmash index`, the index file is named `ecoli.sbt.sbt.json` instead of `ecoli.sbt.json`, which is inconsistent.

## sourmash 3.2.0 error messages

Here's the error message with sourmash 3.2.0 on a zip file:
```
sourmash search ecoli-10.sig ecoli.sbt.zip 

== This is sourmash version 3.2.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: NZ_JSMW01000001.1 Escherichia ... (k=31, DNA)
Error in parsing signature; quitting.
Exception: expected value at line 1 column 1

Cannot open file 'ecoli.sbt.zip'
```

With the unpacked ecoli.sbt.zip --

```
sourmash search ../ecoli-10.sig ecoli.sbt.sbt.json 

== This is sourmash version 3.2.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: NZ_JSMW01000001.1 Escherichia ... (k=31, DNA)
Traceback (most recent call last):
  File "/home/ctbrown/miniconda3/envs/sourmash-test/bin/sourmash", line 11, in <module>
    sys.exit(main())
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/__main__.py", line 14, in main
    return mainmethod(args)
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/cli/search.py", line 62, in main
    return sourmash.commands.search(args)
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/commands.py", line 432, in search
    args.traverse_directory)
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/sourmash_args.py", line 237, in load_dbs_and_sigs
    tree = SBT.load(sbt_or_sigfile, leaf_loader=SigLeaf.load)
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/sbt.py", line 663, in load
    print_version_warning)
  File "/home/ctbrown/miniconda3/envs/sourmash-test/lib/python3.7/site-packages/sourmash/sbt.py", line 820, in _load_v5
    leaves = {int(k): v for (k, v) in info['leaves'].items()}
KeyError: 'leaves'
```
and that seems confusing... so I like the bump to v6.

**Note,** the unpacked index (create as zip -> unpack zip) is `ecoli.sbt.sbt.json` instead of `ecoli.sbt.json`, which seems suboptimal?

## sourmash 2.1.0

It looks like the v5 loader was added after sourmash 2.1.0 (https://github.com/dib-lab/sourmash/commit/7a5d4e646cf4c60b67e4a641f6eda0f5e50190e6).

Attempted loading a zip file & resulting error message:
```
% sourmash search ecoli-10.sig ecoli.sbt.zip 
== This is sourmash version 2.1.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: NZ_JSMW01000001.1 Escherichia ... (k=31, DNA)
Error in parsing signature; quitting.
Exception: lexical error: invalid char in json text.
                                       PK                     (right here) ------^


Cannot open file 'ecoli.sbt.zip'
```

Attempted loading of a new sbt.json version:
```
sourmash search ../ecoli-10.sig ecoli.sbt.sbt.json 
== This is sourmash version 2.1.0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

select query k=31 automatically.
loaded query: NZ_JSMW01000001.1 Escherichia ... (k=31, DNA)
Traceback (most recent call last):
  File "/home/ctbrown/miniconda3/envs/sourmash-test3/bin/sourmash", line 11, in <module>
    sys.exit(main())
  File "/home/ctbrown/miniconda3/envs/sourmash-test3/lib/python3.7/site-packages/sourmash/__main__.py", line 83, in main
    cmd(sys.argv[2:])
  File "/home/ctbrown/miniconda3/envs/sourmash-test3/lib/python3.7/site-packages/sourmash/commands.py", line 875, in search
    args.traverse_directory)
  File "/home/ctbrown/miniconda3/envs/sourmash-test3/lib/python3.7/site-packages/sourmash/sourmash_args.py", line 312, in load_dbs_and_sigs
    tree = SBT.load(sbt_or_sigfile, leaf_loader=SigLeaf.load)
  File "/home/ctbrown/miniconda3/envs/sourmash-test3/lib/python3.7/site-packages/sourmash/sbt.py", line 463, in load
    return loaders[version](jnodes, leaf_loader, dirname, storage,
KeyError: 5
```

That seems fine for past versions. We can maybe document this kind of thing as a version error somewhere in the sourmash docs, or just leave it in the repo...

**BUT** I think we should maybe have a friendlier version error - instead of a KeyError being raised, instead say something like, "The database format in file <blahblah> is not supported this version of sourmash."

**Comment by ctb** [comment]
re zip file sizes, there seems to be no reason to go with -9 at all -- I wouldn't bother even supporting it at the command line, given the below.
```
-rw-rw-r-- 1 ctbrown ctbrown 1471932332 Apr 29 06:52 gtdb-release89-k31.sbt.1.zip
-rw-rw-r-- 1 ctbrown ctbrown 1478489363 Apr 28 10:19 gtdb-release89-k31.sbt.9.zip
```
People can always unzip and manually rezip with higher levels of compression if they really care, right? (Or use the API.)

**Comment by luizirber** [comment]
> for the future, when trying to load an unsupported version of a database, emit an error message like "The database format in file  is not supported this version of sourmash." rather than a KeyError

Changed:
```
try:
    loader = loaders[version]
except KeyError:
    error("The index format in file is not supported this version of sourmash")
    raise SystemExit(1)
```

Better solution: create a new exception type, raise it and catch where appropriate (search/gather CLI, for example). For interactive uses it is horrible to raise `SystemExit(1)`...

I would like to avoid the situation in `load_signatures`, tho: the `try` wrapping around the function makes it hard to debug what problems are happening during development...

> I'm on board with bumping to v6, since the error message for v5 format change is bad :). In the release or commit notes let's make it clear somewhere that no v5 databases should exist in the wild...! We could wait until #925 is merged to do the bump, since that will probably be another incompatibility in created databases, right?

Done. I kept the `_load_v5`, but now defaulting to `v6`. I think the bump can be in this PR, it will be in master but not in any released version (and if you're using `master`, you pay the price of using `master`).

>  we should never be in the situation where older versions of sourmash use khmer to try to load a compressed internal node, right?

Nope. There are no compressed internal nodes before `v6`.

> when creating a zip file with `sourmash index`, the index file is named `ecoli.sbt.sbt.json` instead of `ecoli.sbt.json`, which is inconsistent.

Fixed, I saw it yesterday but had not updated the PR yet.

**Comment by ctb** [comment]
On Wed, Apr 29, 2020 at 08:36:14AM -0700, Luiz Irber wrote:
> Better solution: create a new exception type, raise it and catch where appropriate (search/gather CLI, for example). For interactive uses it is horrible to raise `SystemExit(1)`...

yes. => new issue?

> I would like to avoid the situation in `load_signatures`, tho: the `try` wrapping around the function makes it hard to debug what problems are happening during development...

100% agreed.

> > I'm on board with bumping to v6, since the error message for v5 format change is bad :). In the release or commit notes let's make it clear somewhere that no v5 databases should exist in the wild...! We could wait until #925 is merged to do the bump, since that will probably be another incompatibility in created databases, right?
> 
> Done. I kept the `_load_v5`, but now defaulting to `v6`. I think the bump can be in this PR, it will be in master but not in any released version (and if you're using `master`, you pay the price of using `master`).

yep!

> > when creating a zip file with `sourmash index`, the index file is named `ecoli.sbt.sbt.json` instead of `ecoli.sbt.json`, which is inconsistent.
> 
> Fixed, I saw it yesterday but had not updated the PR yet.

maybe we could add an explicit test for the name? that way we're aware of
when it changes and have that included in a test for diff/blame/changelog
purposes.

**Comment by luizirber** [comment]
> On Wed, Apr 29, 2020 at 08:36:14AM -0700, Luiz Irber wrote: Better solution: create a new exception type, raise it and catch where appropriate (search/gather CLI, for example). For interactive uses it is horrible to raise `SystemExit(1)`...
> yes. => new issue?

this is now raising `IndexNotSupported`, still need to catch it in other places for nicer messages (but the traceback is more informative, at least)

> maybe we could add an explicit test for the name? that way we're aware of when it changes and have that included in a test for diff/blame/changelog purposes.

Added test, and a `sourmash index` test too. I think it is checking too deep into the generated file (in a way that might complicate future changes that keep the test working), but better safe than sorry?

**Comment by ctb** [comment]
Looking pretty fully baked - I have two more files to dig into, but the functionality all seems there! Let me know if you want me to expedite the review.

**Comment by luizirber** [comment]
> Looking pretty fully baked - I have two more files to dig into, but the functionality all seems there! Let me know if you want me to expedite the review.

Almost, it was missing a test for `sourmash index --append`. Now it has one, and the `ZipStorage` can write new entries and avoid duplications (check the [latest commit](https://github.com/dib-lab/sourmash/pull/648/commits/9896db8f273ee3efaf48a92b2662fd3d96317a9d) for untold horrors =])

Ready for final review and merge

File: sourmash/sbt_storage.py [hunk 1]
```diff
@@ -97,6 +107,137 @@ def __exit__(self, type, value, traceback):
         self.tarfile.close()
 
 
+class ZipStorage(Storage):
+
+    def __init__(self, path=None):
+        # TODO: leave it open, or close/open every time?
+
+        if path is None:
+            # TODO: Open a temporary file?
```
Comment (ctb) [inline sourmash/sbt_storage.py:116]: do we want to flag things here (raise an exception, or some such)? or just require path in the constructor?

**Comment by ctb** [comment]
added the issues - now just gotta get the tests passing, I guess :)

File: sourmash/sbt_storage.py (hunk 1 above)
Comment (luizirber) [reply sourmash/sbt_storage.py:116]: I think it would be interesting to do what is happening with the buffer now, where it is kept in memory, and throw a `Warning` if the file is not open? But probably just make it required is easier.

**Comment by ctb** [comment]
very nice work, @luizirber!
